### DISCUSSION TITLE
STM32 board and the U-BLOX module through I2C

### DISCUSSION THREAD

**Discussion by thkinus** [zephyrproject-rtos/zephyr#72698] [Q&A]
Hello everyone,

I am trying to connect the STM32 board(NUCLEO-F767ZI) and the U-BLOX(B78-ADR) module through I2C.

1) I2C communication on STM32CubeIDE without Zephyr
In order to check the I2C connection, I checked I2C communication using STM32CubeIDE without Zephyr. I used the open source below.
(https://github.com/cturvey/RandomNinjaChef/blob/main/STM32_uBlox_I2C.c)
Communication is successful.

2) I2C communication on Zephyr
I tried to I2C communicate with the U-BLOX module using the open source below.
(https://github.com/zephyrproject-rtos/zephyr/pull/46447)
However, an error occurs in the code below.
Is there anything I missed? 

3)  A link to the GPS module  I am using :
: I didn't buy it myself, so I don't have an exact specification.
Below are the product photos and protocol I am referring to below specification documents.

https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf

![image](https://github.com/zephyrproject-rtos/zephyr/assets/127245100/8f572d8e-b979-4873-b9a0-92ace3bbcb3f)

4) I built a image by applying the open source code below only. I didn't apply any modifications.

https://github.com/zephyrproject-rtos/zephyr/pull/46447

5) hardware setup
: I connected as below.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/127245100/0ef746f9-8eac-452f-b5a8-1cd8c7a439d1)

/* i2c_ll_stm32_v2.c */

static int stm32_i2c_msg_write(const struct device *dev, struct i2c_msg *msg,
			uint8_t *next_msg_flags, uint16_t slave)
{
.....

	if (data->current.is_nack || data->current.is_err ||
	    data->current.is_arlo || is_timeout) {
        LOG_ERR("TH : is_nack(%d),is_err(%d),is_arlo(%d),is_timeout(%d)",data->current.is_nack,data->current.is_err,data->current.is_arlo,is_timeout);
		goto error;
	}

	return 0;
error:
....
}

/* Output */

[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : is_nack(0),is_err(0),is_arlo(0),is_timeout(1)
[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : stm32_i2c_msg_write: TIMEOUT
[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : stm32_i2c_transaction(v2) - ret :-5
[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : is_nack(1),is_err(0),is_arlo(0),is_timeout(0)
[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : stm32_i2c_msg_write: NACK
[00:00:00.620,000] <err> i2c_ll_stm32_v2: TH : stm32_i2c_transaction(v2) - ret :-5

**Comment by adri1mart1** [comment]
Any possibility to use a scope on the bus ?

**Reply by thkinus** [reply]
Hello, Thanks for the comment. Sadly... there is no oscilloscope. I think.. I need more additional settings for I2C communication on zephyr, is there a link to refer to?

**Reply by adri1mart1** [reply]
Alright, well could you edit your initial question to add the following information please ?
- A link to the GPS module you are using (I have no idea what is the B78-ADR module :x)
- Are you compiling an existing program of you've create your own ?
- What is your hardware setup ?

Usually I2C is quite straightforward to configure but it really depends on the driver you are using.

**Reply by thkinus** [reply]
Updated.

**Reply by adri1mart1** [reply]
Thanks, quite strange as I cannot find any documentation about the module you are using, it's seems quite old.
I don't see any mistake on the hardware you've shown. 
The first thing I would play with is to update the I2C speed, on your board the value seems to be `clock-frequency = <I2C_BITRATE_FAST>;` (400kHz) which seems to be the maximum value for the M8. You can try to change it to `I2C_BITRATE_STANDARD` (100kHz). 

You saif it's working with the STM32CubeIDE but is it with the exact same hardware setup ? If yes, where is the I2c code initialization so we can compare both ?

**Reply by thkinus** [reply]
Thanks for the comments.

**- The same problem occurs even though set it up as below.**

~/zephyr/boards/st/nucleo_f767zi/nucleo_f767zi.dts
..
&i2c1 {
	pinctrl-0 = <&i2c1_scl_pb8 &i2c1_sda_pb9>;
	pinctrl-names = "default";
	status = "okay";
	clock-frequency = <I2C_BITRATE_STANDARD>;
};


**- Initialization codes**

https://www.st.com/en/embedded-software/stm32cubef7.html#tools-software

Among the examples provided by STM, I2C_TwoBoard_RestartComIT example was referenced.

Q. In the example, various registers are setting. Should the same be implemented in zephyr? 
Is there any doc to refer?

/* main.c */
  /* Configure the system clock to 216 MHz */
  SystemClock_Config();

......
static void SystemClock_Config(void)
{
  RCC_ClkInitTypeDef RCC_ClkInitStruct;
  RCC_OscInitTypeDef RCC_OscInitStruct;

  /* Enable HSE Oscillator and activate PLL with HSE as source */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSE;
  RCC_OscInitStruct.HSEState = RCC_HSE_BYPASS;
  RCC_OscInitStruct.HSIState = RCC_HSI_OFF;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
  RCC_OscInitStruct.PLL.PLLM = 8;
  RCC_OscInitStruct.PLL.PLLN = 432;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = 9;
  RCC_OscInitStruct.PLL.PLLR = 7;

.....

Q. Need set up I2C Timing, is there a way to set it in zephyr?

/* Uncomment this line to use the board as master, if not it is used as slave */
#define I2C_ADDRESS       0x42
#define I2C_TIMING        0x00D00E28  /* (Rise time = 120ns, Fall time = 25ns) */


  /*##-1- Configure the I2C peripheral ######################################*/
  I2cHandle.Instance             = I2Cx;
  I2cHandle.Init.Timing          = I2C_TIMING;
  I2cHandle.Init.OwnAddress1     = I2C_ADDRESS;
  I2cHandle.Init.AddressingMode  = I2C_ADDRESSINGMODE_7BIT;
  I2cHandle.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
  I2cHandle.Init.OwnAddress2     = 0xFF;
  I2cHandle.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
  I2cHandle.Init.NoStretchMode   = I2C_NOSTRETCH_DISABLE;

Thanks
TaeHo

**Reply by adri1mart1** [reply]
Of course you can tweak the code or config files to change a lot of parameters but it really depends on what you need to change. Usually anything anything related to very low level hardware stuff is in dts files, the rest is KConfig options. If you really are on a specific behavior you may eventually need to patch the code yourself but I2C on STM32 to discuss with a UBLOX M8, it's very mature and you should not go for this option at first.

When you build, you have an important file to look at: `build/zephyr/zephyr.dts`. It is the concatenated dts file of your board. Inside, you should see your i2c1 node and all its properties. When you write in your dts 

```
&i2c1 {
pinctrl-0 = <&i2c1_scl_pb8 &i2c1_sda_pb9>;
pinctrl-names = "default";
status = "okay";
clock-frequency = <I2C_BITRATE_STANDARD>;
};
```

The `&` char means you are referring to an existing node call i2c1 and want to update its properties. One node should be compatible with one driver. It you look at the file `build/zephyr/zephyr.dts`, you should see that [i2c1](https://github.com/zephyrproject-rtos/zephyr/blob/68bd4da4c64ca0c19cee0334bee6496ec8cde2b9/dts/arm/st/f7/stm32f7.dtsi#L314) is compatible with this [driver](https://github.com/zephyrproject-rtos/zephyr/blob/68bd4da4c64ca0c19cee0334bee6496ec8cde2b9/drivers/i2c/i2c_ll_stm32.c#L35C23-L35C38). 

Then you can check the code and see where and what is configured. 

Honestly, I would rather use a scope to check the physical i2c signal instead of digging deeper into i2c configuration. I understand you don't have one right now but that's what you need to see what's wrong and adjust either the hardware or the software.

**Comment by erwango** [comment]
Please check if https://github.com/zephyrproject-rtos/zephyr/issues/69290 could help

**Reply by thkinus** [reply]
Thanks for the sharing information.
I don't know if I understand it right...
Do I need to get a new firmware(SDK) from u-blox?

**Reply by erwango** [reply]
@RobMeades should be able to point you in the good direction

**Reply by thkinus** [reply]
@RobMeades I'm experiencing the problem you reported. Could you help me?
